onnxruntime/contrib_ops/cpu/bert/multihead_attention_helper.h

```diff
@@ -27,27 +27,68 @@ Status CheckInputs(const T* query,
+  //  When no packing:
   //     query            (Q)       : (B, S, D)
-  //     key              (K)       : (B, L, D) or (B, N, S*, H)
-  //     value            (V)       : (B, L, D_v) or (B, N, S*, H)
-  //     bias             (Q/K/V)   : (D + D + D_v)
-  // When packed kv is used:
+  //     key              (K)       : (B, L, D)
+  //     value            (V)       : (B, L, D_v)
+  //     bias             (Q/K/V)   : None or (D + D + D_v)
+  //  When it is cross attention (kv cache is not used in this case):
+  //     query            (Q)       : (B, S, D)
+  //     key              (K)       : (B, N, L, H)
+  //     value            (V)       : (B, N, L, H)
+  //     bias             (Q/K/V)   : None
+  //  When packed kv is used:
   //     query            (Q)       : (B, S, D)
   //     key              (K)       : (B, L, N, 2, H)
   //     value            (V)       : None
   //     bias             (Q/K/V)   : None
-  // When packed qkv is used:
-  //     query            (Q)       : (B, L, N, 3, H) or (B, S, 3*D)
+  //  When packed qkv is used (S==L in this case):
+  //     query            (Q)       : (B, S, N, 3, H)
   //     key              (K)       : None
   //     value            (V)       : None
   //     bias             (Q/K/V)   : None or (D + D + D_v)
-
+  //
+  //  Other inputs:
+  //     key_padding_mask (K/V)     : (B) or (3 * B + 2) or (B, T) or (B, S, T)
+  //     relative_position_bias     : (B, N, S, T) or (1, N, S, T)
+  //     past_key                   : (B, N, P, H)
+  //     past_value                 : (B, N, P, H)
+  // ---------------------------------------------------------------
+  // DecoderMaskedMultiHeadAttention inputs (S=1):
+  // ---------------------------------------------------------------
+  //  When no packing:
+  //     query            (Q)       : (B, S, D)
+  //     key              (K)       : (B, L, D)
+  //     value            (V)       : (B, L, D)
+  //  When it is cross attention (kv cache and relative_position_bias are not used, so L==T in this case):
+  //     query            (Q)       : (B, S, D)
+  //     key              (K)       : (B, N, L, H)
+  //     value            (V)       : (B, N, L, H)
+  //  When packed kv is used (S==L in this case):
+  //     query            (Q)       : (B, S, D + D + D_v)
+  //     key              (K)       : None
+  //     value            (V)       : None
+  //
+  //  Other inputs:
+  //     bias             (Q/K/V)   : None or (D + D + D_v)
+  //     key_padding_mask (K/V)     : (B, T)
+  //     relative_position_bias     : (1, N, S, T)
+  //     past_key                   : (B, N, P, H)
+  //     past_value                 : (B, N, P, H)
+  //     past_sequence_length       : scalar (1) when past_present_share_buffer is True, or None
+  //  Remaining inputs (beam_width, cache_indirection) are not checked in the class.
+  // ---------------------------------------------------------------
```

Comment:
I have some very nicely organzied docs about this

https://github.com/microsoft/onnxruntime/blob/b6b29309a529b28e94787edc359f75c9a2207486/onnxruntime/contrib_ops/rocm/bert/batched_gemm_softmax_gemm_permute_pipelines.cuh#L60-L103

The problem is original Attentions are all varies in very small aspect (buffer passing style) but largely serve the same purpose.

Attention, MultiHeadAttention and DecoderMaskedMultiHeadAttention are largely the same. In the long run they should be merged as `BatchAttention`, that is the seq_lens are even in the batch

GroupQueryAttention is not about GQA actually, its unique feature is to support uneven batch (aka, group in GroupQueryAttention has the same meaning as in group_gemm) so in the long run it should be renamed to `GroupAttention`

Another one might deserve an OpKernel is `PagedAttention`.

Just my two cents tho.

Reply:
That's nice. I'll create some documents later.